### PR TITLE
refactor: aws checkForChanges tests for cloudwatch logs use run-serve…

### PR DESCRIPTION
**Pull request by AlinoeDoctari**
This PR begins the work on #8886 it also replaces PR #8857 since the change was required for the refactor.

### CODE REVIEW

**Comment by AlinoeDoctari** [comment]
By respecting the new test interface, I get less coverage than before. I have trouble accessing the coverage report for the diff. Should I try to add more test to improve the coverage?

**Comment by pgrzesik** [comment]
@medikoo Would you be able to review this one? I know you're way more familiar with the scope of these changes. :bow:

File: lib/plugins/aws/deploy/lib/checkForChanges.js [hunk 1]
```diff
@@ -187,7 +187,6 @@ module.exports = {
    */
   async checkLogGroupSubscriptionFilterResourceLimitExceeded() {
     const region = this.provider.getRegion();
-    const cloudWatchLogsSdk = new this.provider.sdk.CloudWatchLogs({ region });
```
Comment (medikoo) [inline lib/plugins/aws/deploy/lib/checkForChanges.js:190]: This PR should be only about tests refactor. Let's not introduce changes to this file

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js [hunk 2]
```diff
@@ -1421,23 +1141,28 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
             headObject: async (params) => generateMatchingHeadObjectResponse(serverless, params),
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1141]: `generateMatchingListObjectsResponse` and `generateMatchingHeadObjectResponse` functions has some TODO parts not addressed.

To have this test fully complete either we have to update those functions, or if there's no need to rely on them in this case, we can simply put payload inline.

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js [hunk 3]
```diff
@@ -1421,23 +1141,28 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
...
-              // TODO: Ensure same ARN as lambda on which it is configured
+            deleteSubscriptionFilter: deleteStub,
+            describeSubscriptionFilters: async () => {
+              const functionName = serverless.service.getFunction('fn1').name;
+              return {
+                subscriptionFilters: [
+                  {
+                    // destinationArn `arn:{partition}:lambda:{region}:{accountId}:function:{functionName}`
+                    // filterName {stack name}-{logical id}-{random alphanumeric characters}
+                    filterName: 'teststack-Fn1LogsSubscriptionFilterCloudWatchLog1-xxxxx',
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1152]: It'll be nice to generate this name with help of `naming` utils. We have access to stack name via `naming.getStackName()`, we can get _logical id_ via `naming.getCloudWatchLogLogicalId()`

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js [hunk 4]
```diff
@@ -1503,22 +1234,47 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
...
+                    {
+                      filterName: 'teststack-Fn1LogsSubscriptionFilterCloudWatchLog1-xxxxx',
+                      destinationArn: `arn:aws:lambda:us-east-1:999999999999:function:${
+                        serverless.service.getFunction('fn1').name
+                      }`,
+                    },
+                  ],
+                })
+              )
+              .onSecondCall()
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1253]: Why do we configure results for two calls here?

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js [hunk 5]
```diff
@@ -1503,22 +1234,47 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
...
-              // TODO: Ensure same ARN as lambda on which it is configured, but tweak index for one of the filters
-            },
-            // TODO: Configure stub
-            deleteSubscriptionFilter: null,
+            deleteSubscriptionFilter: deleteStub,
+            describeSubscriptionFilters: sandbox
+              .stub()
+              .onFirstCall()
+              .callsFake(() =>
+                Promise.resolve({
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1242]: Let's use async function instead

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js [hunk 6]
```diff
@@ -1503,22 +1234,47 @@ describe('test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js', () => {
...
+        logGroupName: 'someLogGroupName2',
+        filterName: 'teststack-Fn1LogsSubscriptionFilterCloudWatchLog3-xxxxx',
+      });
     });
 
     it('should not attempt to delete and add filter in context of custom partition', async () => {
-      // Replaces:
-      // https://github.com/serverless/serverless/blob/61dd3bde8d17cdd995fdd27259a689d12bee1e42/test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js#L785-L814
-
+      const deleteStub = sandbox.stub();
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1277]: If see correctly, this test is simply to confirm that we recognized custom partitions (as `aws-us-gov`), let's update it's title to _"should recognize custom partition"_ (current title is misleading, as it suggests that we simply never delete filters if it's set in custom partition, that's not true)

**Review by medikoo — CHANGES_REQUESTED** [review]
@AlinoeDoctari thank you! It looks really good, I have just few suggestions, please check my comments

**Comment by AlinoeDoctari** [comment]
Thanks a lot for the review :bowing_man: I focused on understanding the run-serverless part and clearly missed a lot of things, sorry. I'll review all that and try to come up with something a bit better :smile:

File: lib/plugins/aws/deploy/lib/checkForChanges.js (hunk 1 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: I had a look again, and I'm not sure how I should stub the sdk.CloudwatchLogs with runServerless. From what I understand, awsRequestStubMap only works for to replace request in the provider. Is there a way to stub the provider.sdk without doing it "manually"?

Comment (medikoo) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: Indeed @AlinoeDoctari, I totally missed that. Ideally if stub somehow stubs `aws-sdk` directly, still to not make a dependency on that here, let's keep this refactor here as you proposed initially.

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js (hunk 2 above)
Comment (AlinoeDoctari) [reply test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1158]: Function are now completed.

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js (hunk 3 above)
Comment (AlinoeDoctari) [reply test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js]: I think I've come up with the expected expression here, let me know if you had something else in mind.

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js (hunk 5 above)
Comment (AlinoeDoctari) [reply test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js]: No more promise there :+1:

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js (hunk 4 above)
Comment (AlinoeDoctari) [reply test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1284]: We map the fix function on the logs group in the checkForChanges implementation, as we have two log groups there, we get two calls.

File: test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js (hunk 6 above)
Comment (AlinoeDoctari) [reply test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js:1315]: Title changed.

**Comment by AlinoeDoctari** [comment]
I got a bit  busy so it took more time than expected but it should now be ready for a second review @medikoo :smile:

**Review by medikoo — APPROVED** [review]
@AlinoeDoctari that looks great! Thank you

**Pull request merged**